### PR TITLE
Ruby Course (various): Link updates as per Issue 28290

### PR DESCRIPTION
**Pull request by MidlifeMechanicus**
## Because

Links in our lessons should be sufficiently descriptive to improve accessibility. As per the [layout style guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md#links), link text should both concisely describe its purpose (or where the link will redirect a user), and must not include the words this nor here.

Not all links currently meet the above standard so an editing pass was organized.

## This PR

- Update link descriptions in Test Driven Development
- Update link descriptions in Project: Connect Four
- Update link descriptions in Project: Ruby Final Project

## Issue

Related to #28290

## Pull Request Requirements

-   [x] I have thoroughly read and understand [The Odin Project curriculum contributing guide](https://github.com/TheOdinProject/curriculum/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`
-   [x] The `Because` section summarizes the reason for this PR
-   [x] The `This PR` section has a bullet point list describing the changes in this PR
-   [x] If this PR addresses an open issue, it is linked in the `Issue` section
-   [x] If any lesson files are included in this PR, they have been previewed with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure it is formatted correctly
-   [x] If any lesson files are included in this PR, they follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

### CODE REVIEW

**Comment by MidlifeMechanicus** [comment]
The remaining linting issue seems to want an Assignment heading at project_connect_four.md:45. That would be the very last line of the lesson, so I am assuming this linting flag has been raised in error.

If you need anything further from me, though, just let me know.

File: ruby/conclusion/project_ruby_final.md [hunk 1]
```diff
@@ -31,7 +32,7 @@ This is a great project to have as a part of your portfolio going forward becaus
 
 This section contains helpful links to related content. It isn't required, so consider it supplemental.
 
-- [Illustrated rules of Chess](http://www.chessvariants.org/d.chess/chess.html)
-- [Chess notation for humans and computers](https://en.wikipedia.org/wiki/Chess_notation)
+- Chess Variant's [illustrated rules](http://www.chessvariants.org/d.chess/chess.html) of Chess.
+- Wikipedia's entry on [Chess notation for humans and computers](https://en.wikipedia.org/wiki/Chess_notation)
 - If you're feeling stuck, you should visit the community Discord!
-- Once you have completed the project, check out this [discussion](http://rubyquiz.com/quiz35.html) of the problem from RubyQuiz. **Warning**: this resource contains a blueprint for how to complete this project. That means, if you decide to peruse it, you won't get the opportunity to dig into this problem.
+- Once you have completed the project, check out the RubyQuiz [discussion](http://rubyquiz.com/quiz35.html) of the problem. **Warning**: this resource contains a blueprint for how to complete this project. That means, if you decide to peruse it, you won't get the opportunity to dig into this problem.
```
Comment (MaoShizhong) [inline ruby/conclusion/project_ruby_final.md:38]: The link text itself has to be descriptive enough on its own. even if you change the surrounding text to say `the RubyQuiz`, it's not part of the link text and navigating by link, a screenreader would just say "discussion - link", which isn't sufficiently descriptive of either where it leads or what it leads to.

```suggestion
- Once you have completed the project, check out this [discussion on writing chess in Ruby](http://rubyquiz.com/quiz35.html) from RubyQuiz. **Warning**: this resource contains a blueprint for how to complete this project. That means, if you decide to peruse it, you won't get the opportunity to dig into this problem.
```

File: ruby/testing_ruby_with_rspec/project_connect_four.md [hunk 2]
```diff
@@ -7,20 +7,21 @@ You still may feel shaky on RSpec at this point (which is totally normal), so le
...
-</div>
+      1. Test each of your critical methods to make sure they function properly and handle edge cases.
+      1. Use mocks/doubles to isolate methods to make sure that they're sending back the right outputs.
 
+</div>
 
 ### Project: TDD Connect Four
 
-Hopefully everyone has played Connect Four at some point (if not, see the [Wikipedia page](http://en.wikipedia.org/wiki/Connect_Four)).  It's a basic game where each player takes turns dropping pieces into the cage.  Players win if they manage to get 4 of their pieces consecutively in a row, column, or along a diagonal.
+Hopefully everyone has played Connect Four at some point (if not, see the Connect Four [Wikipedia page](http://en.wikipedia.org/wiki/Connect_Four)).  It's a basic game where each player takes turns dropping pieces into the cage.  Players win if they manage to get 4 of their pieces consecutively in a row, column, or along a diagonal.
```
Comment (MaoShizhong) [inline ruby/testing_ruby_with_rspec/project_connect_four.md:22]: Same here, "Wikipedia page - link" isn't sufficiently accessible and would force the user to have to have the whole line read out to try and get the context for what the link contains or leads to. In the suggestion, "Connect Four - link" is more descriptive because it leads directly to a page titled "Connect Four".

```suggestion
Hopefully most people will have played [Connect Four](http://en.wikipedia.org/wiki/Connect_Four) at some point. It's a basic game where each player takes turns dropping pieces into the cage. Players win if they manage to get 4 of their pieces consecutively in a row, column, or along a diagonal.
```

File: ruby/testing_ruby_with_rspec/project_connect_four.md [hunk 3]
```diff
@@ -29,12 +30,15 @@ Only write exactly enough code to make your test pass.  Oftentimes, you'll end u
...
 ### Additional resources
-This section contains helpful links to other content. It isn't required, so consider it supplemental.
 
-* [RSpec Mock example SO question](http://stackoverflow.com/questions/3622604/rspec-mock-object-example)
-* [Helpful video from Sandi Metz](https://www.youtube.com/watch?v=URSWYvyc42M)
-* [Tutorials Point RSpec Writing Specs](https://www.tutorialspoint.com/rspec/rspec_writing_specs.htm)
+This section contains helpful links to related content. It isn't required, so consider it supplemental.
+
+- Stack Overflow has an [RSpec Mock Object example](http://stackoverflow.com/questions/3622604/rspec-mock-object-example).
+- Youtube has a [helpful video](https://www.youtube.com/watch?v=URSWYvyc42M) by Sandi Metz about Ruby testing.
```
Comment (MaoShizhong) [inline ruby/testing_ruby_with_rspec/project_connect_four.md:43]:
```suggestion
- Youtube has a [helpful video by Sandi Metz about Ruby testing](https://www.youtube.com/watch?v=URSWYvyc42M).
```

File: ruby/testing_ruby_with_rspec/test_driven_development.md [hunk 4]
```diff
@@ -13,47 +13,49 @@ This section contains a general overview of topics that you will learn in this l
...
+As outlined in [Wikipedia](https://en.wikipedia.org/wiki/Test-driven_development), Test-Driven Development (TDD) is a process and technique of software development that relies on the repetition of a very short and specific development cycle. In each cycle, requirements (i.e., what you want your code to do) are turned into specific test cases first. These requirements could be anything from an entire feature that requires end-to-end (E2E) testing, such as tests that cover a user logging into your website successfully AND unsuccessfully, to a new Ruby class you've devised, for which unit tests might suffice. Either way, the test suite for these requirements fail initially, since actual code hasn't been written yet.
```
Comment (MaoShizhong) [inline ruby/testing_ruby_with_rspec/test_driven_development.md:16]: Can this be reverted back to the original? Same reasons as above regarding the new link text, and the original is sufficiently accessible.

File: ruby/testing_ruby_with_rspec/test_driven_development.md [hunk 5]
```diff
@@ -13,47 +13,49 @@ This section contains a general overview of topics that you will learn in this l
...
+Even if TDD intuitively seems like good practice, its utility is hotly debated even today.
 
-No one is saying that tests aren't useful. Testing at [all levels](https://thoughtbot.com/blog/rails-test-types-and-the-testing-pyramid) is *crucial*. Testing ensures that your application is working as intended. More importantly, and perhaps less obviously, testing ensures developer confidence when it comes time to change existing code. Adding new features and dependencies increases the overall instability of your application, and a single misstep can have a cascading, crashing effect in faraway, unintended locations. Testing confirms that what you add works with what you already have. In a sense, it provides a safety net that leaves your application more open to change, and gives it room to grow and evolve as painlessly as possible.
+No one is saying that tests aren't useful. Testing at all levels is *crucial*. Joël Quenneville's article on the Rails [Testing Pyramid](https://thoughtbot.com/blog/rails-test-types-and-the-testing-pyramid) provides an overview into the different types and levels of testing that Rails developers routinely use.
```
Comment (MaoShizhong) [inline ruby/testing_ruby_with_rspec/test_driven_development.md:26]: Good catch! Missed that from my spreadsheet :pray:

File: ruby/testing_ruby_with_rspec/test_driven_development.md [hunk 6]
```diff
@@ -13,47 +13,49 @@ This section contains a general overview of topics that you will learn in this l
...
+- [What is the difference between unit and integration tests?](https://thoughtbot.com/blog/rails-test-types-and-the-testing-pyramid)
+- [How does TDD help in code refactoring?](https://www.madetech.com/blog/9-benefits-of-test-driven-development)
 
 ### Additional resources
 
 This section contains helpful links to related content. It isn't required, so consider it supplemental.
 
-- [This video](https://www.youtube.com/watch?v=PCEHRFHKZSk) that provides counter arguments for DHH’s [“TDD is Dead. Long Live Testing”](https://dhh.dk/2014/tdd-is-dead-long-live-testing.html)
-- [This series of talks](https://martinfowler.com/articles/is-tdd-dead/) that provides an even more rounded take on TDD.
+- Youtube has [a video](https://www.youtube.com/watch?v=PCEHRFHKZSk) that provides counter arguments for DHH’s [“TDD is Dead. Long Live Testing”](https://dhh.dk/2014/tdd-is-dead-long-live-testing.html).
```
Comment (MaoShizhong) [inline ruby/testing_ruby_with_rspec/test_driven_development.md:60]: Same reasons as earlier
```suggestion
- This video that provides [counter arguments for DHH's “TDD is Dead. Long Live Testing” article](https://www.youtube.com/watch?v=PCEHRFHKZSk) from the assignment.
```

File: ruby/testing_ruby_with_rspec/test_driven_development.md [hunk 7]
```diff
@@ -13,47 +13,49 @@ This section contains a general overview of topics that you will learn in this l
...
+- Martin Fowler has [a series of talks](https://martinfowler.com/articles/is-tdd-dead/) that provides an even more rounded take on TDD.
```
Comment (MaoShizhong) [inline ruby/testing_ruby_with_rspec/test_driven_development.md:61]:
```suggestion
- For a more rounded taken on TDD, watch or listen to this [series of talks on the pros and cons of TDD](https://martinfowler.com/articles/is-tdd-dead/).
```
